sdk/metric/integrator/simple/simple.go

```diff
@@ -68,97 +84,271 @@ type (
+		// allocated in the `!ok` branch if this is stateful
+		// PrecomputedSum instrument (in which case the exporter
+		// is requesting a delta so we allocate it up front),
+		// and it would be allocated in this block when multiple
+		// accumulators are used and the first condition is not
+		// met.
+		b.AggregationSelector.AggregatorFor(desc, &value.delta)
+	}
+	if value.current != value.delta {
+		// If the current and delta Aggregators are not the same it
+		// implies that multiple Accumulators were used.  The first
+		// Accumulation seen for a given stateKey will return in
+		// one of the cases above after assigning `value.current
+		// = agg` (i.e., after taking a reference to the
+		// Accumulator's Aggregator).
+		//
+		// The second time through this branch copies the
+		// Accumulator's Aggregator into `value.delta` and sets
+		// `value.current` appropriately to avoid this branch if
+		// a third Accumulator is used.
+		err := value.current.SynchronizedCopy(value.delta, desc)
+		if err != nil {
+			return err
+		}
+		value.current = value.delta
+	}
+	// The two statements above ensures that `value.current` refers
+	// to `value.delta` and not to an Accumulator's Aggregator.  Now
+	// combine this Accumulation with the prior Accumulation.
+	return value.delta.Merge(agg, desc)
 }
 
+// CheckpointSet returns the associated CheckpointSet.  Use the
+// CheckpointSet Locker interface to synchronize access to this
+// object.  The CheckpointSet.ForEach() method cannot be called
+// concurrently with Process().
 func (b *Integrator) CheckpointSet() export.CheckpointSet {
-	return &b.batch
+	return &b.state
 }
 
+// StartCollection signals to the Integrator one or more Accumulators
+// will begin calling Process() calls during collection.
 func (b *Integrator) StartCollection() {
 	if b.startedCollection != 0 {
 		b.intervalStart = b.intervalEnd
 	}
 	b.startedCollection++
-	if !b.stateful {
-		b.batch.values = map[batchKey]batchValue{}
-	}
 }
 
+// FinishCollection signals to the Integrator that a complete
+// collection has finished and that ForEach will be called to access
+// the CheckpointSet.
 func (b *Integrator) FinishCollection() error {
-	b.finishedCollection++
 	b.intervalEnd = time.Now()
-	if b.startedCollection != b.finishedCollection {
+	if b.startedCollection != b.finishedCollection+1 {
 		return ErrInconsistentState
 	}
+	defer func() { b.finishedCollection++ }()
+
+	for key, value := range b.values {
+		mkind := key.descriptor.MetricKind()
+
+		if !value.stateful {
+			if value.updated != b.finishedCollection {
+				delete(b.values, key)
+			}
+			continue
+		}
+
+		// Update Aggregator state to support exporting either a
+		// delta or a cumulative aggregation.
+		var err error
+		if mkind.PrecomputedSum() {
+			// delta_value = current_cumulative_value - previous_cumulative_value
+			if subt, ok := value.current.(export.Subtractor); ok {
+				err = subt.Subtract(value.cumulative, value.delta, key.descriptor)
```

Comment:
I did this and added a bit more comment. I considered what to name such an in-place operation, and the exercise left me feeling like it would add more confusion than it would relieve. There's not a good way to say "T <- S - T".  I'd be happy to do this in a future PR.

Reply:
Or even better would it be possible to just return the result so this could be written as:
```go
value.delta = currentSubtractor.Subtract(value.cumulative, key.descriptor)
```
(I seem to remember thinking that's not possible but can't remember why)